sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -337,11 +340,19 @@ trait DivModLike extends BinaryArithmetic {
     } else {
       s"${eval2.value} == 0"
     }
-    val javaType = CodeGenerator.javaType(dataType)
+    val isIntegralDiv = this.isInstanceOf[IntegralDivide]
+    // From SPARK-16323 IntegralDivision returns Long data type
+    val javaType = if (isIntegralDiv) JAVA_LONG else CodeGenerator.javaType(dataType)
+    val operandJavaType = if (isIntegralDiv) operandsDataType match {
+      case _: IntegerType => JAVA_LONG
```

Comment:
Have you tried to fix this issue in `TypeCoercion`? It seems that is a more suitable place to fix it though.

Reply:
Yeah, let's handle it there if other DBMSes allow.

Reply:
I ran the `select -2147483648 / -1 ` query in some of the databases, results are as below

MySQL => 2147483648

PostgresSQl9.6 => Out of Integer Range exception

mariaDB => 2147483648

MS Sql Server 2017 => 2147483648

Reply:
Behaviour is mixed w.r.t other DBMSes
since `IntegralDivide` returns Long, shall I cast to Long in `TypeCoercion` ?

Reply:
Okay, allowing it seems making sense. It will also keep the compatibility.

Reply:
okay will update the PR with `TypeCoercion` implementation